### PR TITLE
Consistent initializer parameter names for ActionDispatch::Http::UploadedFile

### PR DESCRIPTION
**Pull request by thepoho**
ActionDispatch::Http::UploadedFile.new accepts some parameters in
it's initializer for content-type and headers. These are named
`type` and `head`, which is unintuitive.  This PR allows the object
to be initialized with `content_type` and `headers` to bring
consistency with the name of the actual attributes. It also
maintains backwards compatibility.

    ActionDispatch::Http::UploadedFile.new({
      tempfile:     tempfile,
      content_type: 'image/svg+xml',
      headers:      'some_headers',
      filename:     'filename.svg',
    })

### Summary

I came across a gotcha when initializing `ActionDispatch::Http::UploadedFile`.  The documentation states the attributes are `content_type` and `headers` however when initializing it expects `type` and `head` respectively, which is un-intuitive.  This PR allows the correct parameter names when initializing the class, and maintains backwards compatibility with the old names.

### CODE REVIEW

File: actionpack/lib/action_dispatch/http/upload.rb [hunk 1]
```diff
@@ -41,8 +41,8 @@ def initialize(hash) # :nodoc:
           @original_filename = nil
         end
 
-        @content_type      = hash[:type]
-        @headers           = hash[:head]
+        @content_type = hash[:type] || hash[:content_type]
+        @headers      = hash[:head] || hash[:headers]
```
Comment (eileencodes) [inline actionpack/lib/action_dispatch/http/upload.rb]: What do you think about deprecating the old, unintuitive ones?

File: actionpack/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,18 @@
...
+
+    This keeps the constructor consistent with the names of the parameters. 
+    Also maintains the legacy parameters of `type` and `head` for backwards compatibility
+
+      ActionDispatch::Http::UploadedFile.new({
+        tempfile:     tempfile,
+        content_type: 'image/svg+xml',
+        headers:      'some_headers',
+        filename:     'filename.svg',
+      })
```
Comment (kaspth) [inline actionpack/CHANGELOG.md]: Let's leave the braces to the high schoolers, we don't need them here. 😄

File: actionpack/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,18 @@
...
+
+    *Peter Hofmann*
+
+
```
Comment (kaspth) [inline actionpack/CHANGELOG.md]: One newline too much.

File: actionpack/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,18 @@
+*   `ActionDispatch::Http::UploadedFile` now accepts `content_type` and `headers` when initializing
+
+    This keeps the constructor consistent with the names of the parameters. 
+    Also maintains the legacy parameters of `type` and `head` for backwards compatibility
+
+      ActionDispatch::Http::UploadedFile.new({
```
Comment (kaspth) [inline actionpack/CHANGELOG.md]: Code is indented by 2 more spaces here.

File: actionpack/lib/action_dispatch/http/upload.rb (hunk 1 above)
Comment (kaspth) [reply actionpack/lib/action_dispatch/http/upload.rb]: Might also be nice to switch to keyword arguments while we're here. Saves us the ArgumentError when missing `:tempfile`.

**Review by kaspth — COMMENTED** [review]
Neat!

File: actionpack/lib/action_dispatch/http/upload.rb (hunk 1 above)
Comment (sikachu) [reply actionpack/lib/action_dispatch/http/upload.rb]: > What would the process be for creating the deprecation? Is there a guide?

Should be as simple as:

```ruby
# before line 44
if hash.key?(:type)
  ActiveSupport::Deprecation.warn("...")
end

if hash.key?(:head)
  # ...
```

Note that you'll need to add `require "active_support/deprecation` on the top of the file too. You can try grep the project for `Deprecation.warn` to see examples.

---

> what version should we target for deprecation?

I think it's safe to say _this will be removed in the next major version of Rails_, which essentially mean 6.0

---

And yes, let's change to kwargs when we're at it. 💯

**Comment by oniofchaos** [comment]
@thepoho would you like any assistance in getting this across the finish line? Great idea on the naming improvement for clarity.

**Comment by thepoho** [comment]
Hey sorry it's been a while. I ran into an issue using named parameters. It seems there is a bonus parameter coming in (which is then discarded) from Rack Test in certain circumstances. Will get back to it in the next few days.

**Comment by thepoho** [comment]
I've added the deprecation warnings.  I'm thinking kwargs should be tackled at the time of deprecation to avoid having two sets of similar named args for this method.  There were also some issues with additional arguments being passed in and never used (which would break with kwargs) which needs looking into further at the time of moving to kwargs

**Comment by thepoho** [comment]
@oniofchaos @kaspth @rafaelfranca Anything left to do here?  Turns out this is going stale...

File: actionpack/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,17 @@
+*   `ActionDispatch::Http::UploadedFile` now accepts `content_type` and `headers` when initializing
```
Comment (kaspth) [inline actionpack/CHANGELOG.md]: Screen Shot 2019-12-18 at 23 08 14

Does the `100644 → 100755` mean that we're changing file permissions? If so, let's not. Seems like they're on every file too.

File: actionpack/lib/action_dispatch/http/upload.rb [hunk 6]
```diff
@@ -41,8 +42,21 @@ def initialize(hash) # :nodoc:
...
-        @headers           = hash[:head]
+        if hash.key?(:type)
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
+            Initialising with key `type` is deprecated and will be removed in Rails 6.0.
+            Please use key `content_type` instead.
+          MSG
+        end
+        if hash.key?(:head)
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
+            Initialising with key `head` is deprecated and will be removed in Rails 6.0.
```
Comment (kaspth) [inline actionpack/lib/action_dispatch/http/upload.rb]: Needs to be Rails 6.1 now. Same in the other one.

File: actionpack/lib/action_dispatch/http/upload.rb [hunk 7]
```diff
@@ -1,5 +1,6 @@
 # frozen_string_literal: true
 
+require "active_support/deprecation"
```
Comment (kaspth) [inline actionpack/lib/action_dispatch/http/upload.rb]: Newline after this.

File: actionpack/test/dispatch/uploaded_file_test.rb [hunk 8]
```diff
@@ -33,15 +33,29 @@ def test_filename_should_always_be_in_utf_8
     end
 
     def test_content_type
-      uf = Http::UploadedFile.new(type: "foo", tempfile: Object.new)
+      uf = Http::UploadedFile.new(content_type: "foo", tempfile: Object.new)
       assert_equal "foo", uf.content_type
     end
 
+    def test_content_type_legacy
+      ActiveSupport::Deprecation.silence do
```
Comment (kaspth) [inline actionpack/test/dispatch/uploaded_file_test.rb]: `assert_deprecated` here and in the other one?

File: actionpack/CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,17 @@
...
+
+    This keeps the constructor consistent with the names of the parameters.
+    Also maintains the legacy parameters of `type` and `head` for backwards compatibility
```
Comment (kaspth) [inline actionpack/CHANGELOG.md]: `Also deprecates the…`, sentence needs a period at the end too.

File: actionpack/CHANGELOG.md (hunk 5 above)
Comment (kaspth) [reply actionpack/CHANGELOG.md]: Also needs a period at the end of the title to fit with our changelog style.

**Review by kaspth — COMMENTED** [review]
Thanks for your patience here, looks like this is close. Can you squash your commits too?

File: actionpack/lib/action_dispatch/http/upload.rb (hunk 6 above)
Comment (eugeneius) [reply actionpack/lib/action_dispatch/http/upload.rb]: This change would be released in 6.1, so the deprecation should mention 6.2, right?

Comment (kaspth) [reply actionpack/lib/action_dispatch/http/upload.rb]: Hey, let me enjoy my off-by-one errors in peaceful ignorance, please 😂

You're right, it's 6.2, thanks for catching it! ❤️

**Pull request closed without merging**